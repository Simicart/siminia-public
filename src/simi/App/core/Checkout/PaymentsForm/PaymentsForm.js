import React, { useCallback, useState } from 'react';
import { Form } from 'informed';
import { array, shape, string, object } from 'prop-types';

import { mergeClasses } from 'src/classify';
import defaultClasses from './PaymentsForm.css';
import PaymentsFormItems from '../components/paymentsFormItems';

/**
 * A wrapper around the payment form. This component's purpose is to maintain
 * the submission state as well as prepare/set initial values.
 */
const PaymentsForm = props => {
    const { initialValues } = props;
    const classes = mergeClasses(defaultClasses, props.classes);

    const [isSubmitting, setIsSubmitting] = useState(false);

    const handleSubmit = useCallback(() => {
        setIsSubmitting(true);
    }, [setIsSubmitting]);

    const formChildrenProps = {
        ...props,
        classes,
        isSubmitting,
        setIsSubmitting
    };


    return (
        <Form
            className={classes.root}
            initialValues={initialValues}
            onSubmit={handleSubmit}
        >
            <PaymentsFormItems {...formChildrenProps} />
        </Form>
    );
};

PaymentsForm.propTypes = {
    classes: shape({
        root: string
    }),
    initialValues: object,
    paymentMethods: array
};

PaymentsForm.defaultProps = {
    initialValues: {}
};

export default PaymentsForm;
