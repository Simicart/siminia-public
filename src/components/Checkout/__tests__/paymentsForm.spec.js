import React from 'react';
import testRenderer from 'react-test-renderer';

import PaymentsForm from '../paymentsForm';
import { Form } from 'informed';
import BraintreeDropin from '../braintreeDropin';

jest.mock('src/classify');

const mockCancel = jest.fn();
const mockSubmit = jest.fn();
const defaultProps = {
    cancel: mockCancel,
    submit: mockSubmit
};

beforeEach(() => {
    mockCancel.mockReset();
    mockSubmit.mockReset();
});
test('renders a PaymentsForm component', () => {
    const component = testRenderer.create(<PaymentsForm {...defaultProps} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('sets addresses_same true if no initial form values', () => {
    const props = {
        ...defaultProps,
        initialValues: {}
    };
    const component = testRenderer.create(<PaymentsForm {...props} />);

    expect(
        component.root.findByType(Form).props.initialValues.addresses_same
    ).toEqual(true);
});

test('sets addresses_same true if initialValues.sameAsShippingAddress is true', () => {
    const props = {
        ...defaultProps,
        initialValues: {
            sameAsShippingAddress: true
        }
    };
    const component = testRenderer.create(<PaymentsForm {...props} />);

    expect(
        component.root.findByType(Form).props.initialValues.addresses_same
    ).toEqual(true);
});

test('sets addresses_same false if initialValues.sameAsShippingAddress is false', () => {
    const props = {
        ...defaultProps,
        initialValues: {
            sameAsShippingAddress: false
        }
    };
    const component = testRenderer.create(<PaymentsForm {...props} />);

    expect(
        component.root.findByType(Form).props.initialValues.addresses_same
    ).toEqual(false);
});

test('renders billing address fields if addresses_same checkbox unchecked', () => {
    const props = {
        ...defaultProps,
        initialValues: {
            sameAsShippingAddress: false
        }
    };
    const component = testRenderer.create(<PaymentsForm {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(component.root.findByProps({ className: 'street0' })).toBeTruthy();
});

test('setPaymentNonce function submits billing address and payment method', () => {
    const props = {
        ...defaultProps,
        initialValues: {
            sameAsShippingAddress: true
        }
    };
    const component = testRenderer.create(<PaymentsForm {...props} />);

    const testNonce = 'testData';
    component.root.children[0].instance.setPaymentNonce(testNonce);

    expect(mockSubmit).toHaveBeenCalledWith({
        billingAddress: {
            sameAsShippingAddress: true
        },
        paymentMethod: {
            code: 'braintree',
            data: testNonce
        }
    });
});

test('setPaymentNonce function gets billing address from form if not same as shipping', () => {
    const props = {
        ...defaultProps,
        initialValues: {
            sameAsShippingAddress: false
        }
    };
    const component = testRenderer.create(<PaymentsForm {...props} />);

    const testNonce = 'testData';
    component.root.children[0].instance.setPaymentNonce(testNonce);

    expect(mockSubmit).toHaveBeenCalledWith({
        billingAddress: {
            city: undefined,
            postcode: undefined,
            region_code: undefined,
            street: undefined
        },
        paymentMethod: {
            code: 'braintree',
            data: testNonce
        }
    });
});

test('cancel instance function calls props cancel function', () => {
    const component = testRenderer.create(<PaymentsForm {...defaultProps} />);
    component.root.children[0].instance.cancel();

    expect(mockCancel).toHaveBeenCalled();
});

test('submit instance function sets isRequestingPaymentNonce true in state', () => {
    const component = testRenderer.create(<PaymentsForm {...defaultProps} />);

    component.root.children[0].instance.submit();

    expect(
        component.root.findByType(BraintreeDropin).props
            .isRequestingPaymentNonce
    ).toEqual(true);
});

test('cancelPaymentNonceRequest instance function sets isRequestingPaymentNonce false in state', () => {
    const component = testRenderer.create(<PaymentsForm {...defaultProps} />);

    component.root.children[0].instance.cancelPaymentNonceRequest();

    expect(
        component.root.findByType(BraintreeDropin).props
            .isRequestingPaymentNonce
    ).toEqual(false);
});
