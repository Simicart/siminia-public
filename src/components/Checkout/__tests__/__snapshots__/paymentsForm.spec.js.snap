// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a PaymentsForm component 1`] = `
<form
  className="root"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="body"
  >
    <h2
      className="heading"
    >
      Billing Information
    </h2>
    <div
      className="braintree"
    >
      <div
        className="root"
      >
        <div
          id="braintree-dropin-container"
        />
      </div>
    </div>
    <div
      className="address_check"
    >
      <label
        className="root"
      >
        <span
          className="icon"
        >
          <span
            className="root"
          >
            <svg
              fill="none"
              height={18}
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width={18}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </span>
        </span>
        <input
          checked={true}
          className="input"
          name="addresses_same"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="label"
        >
          Billing address same as shipping address
        </span>
      </label>
      <p
        className="root"
      />
    </div>
  </div>
  <div
    className="footer"
  >
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        Cancel
      </span>
    </button>
    <button
      className="button"
      type="submit"
    >
      <span
        className="content"
      >
        Use Card
      </span>
    </button>
  </div>
</form>
`;

exports[`renders billing address fields if addresses_same checkbox unchecked 1`] = `
<form
  className="root"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="body"
  >
    <h2
      className="heading"
    >
      Billing Information
    </h2>
    <div
      className="braintree"
    >
      <div
        className="root"
      >
        <div
          id="braintree-dropin-container"
        />
      </div>
    </div>
    <div
      className="address_check"
    >
      <label
        className="root"
      >
        <span
          className="icon"
        />
        <input
          checked={false}
          className="input"
          name="addresses_same"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="label"
        >
          Billing address same as shipping address
        </span>
      </label>
      <p
        className="root"
      />
    </div>
    <div
      className="street0"
    >
      <div
        className="root"
      >
        <span
          className="label"
        >
          Street
        </span>
        <span
          className="root"
          style={
            Object {
              "--iconsAfter": 0,
              "--iconsBefore": 0,
            }
          }
        >
          <span
            className="input"
          >
            <input
              className="input"
              id="street0"
              name="street[0]"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            />
          </span>
          <span
            className="before"
          />
          <span
            className="after"
          />
        </span>
        <p
          className="root"
        />
      </div>
    </div>
    <div
      className="city"
    >
      <div
        className="root"
      >
        <span
          className="label"
        >
          City
        </span>
        <span
          className="root"
          style={
            Object {
              "--iconsAfter": 0,
              "--iconsBefore": 0,
            }
          }
        >
          <span
            className="input"
          >
            <input
              className="input"
              id="city"
              name="city"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            />
          </span>
          <span
            className="before"
          />
          <span
            className="after"
          />
        </span>
        <p
          className="root"
        />
      </div>
    </div>
    <div
      className="region_code"
    >
      <div
        className="root"
      >
        <span
          className="label"
        >
          State
        </span>
        <span
          className="root"
          style={
            Object {
              "--iconsAfter": 0,
              "--iconsBefore": 0,
            }
          }
        >
          <span
            className="input"
          >
            <input
              className="input"
              id="region_code"
              name="region_code"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            />
          </span>
          <span
            className="before"
          />
          <span
            className="after"
          />
        </span>
        <p
          className="root"
        />
      </div>
    </div>
    <div
      className="postcode"
    >
      <div
        className="root"
      >
        <span
          className="label"
        >
          ZIP
        </span>
        <span
          className="root"
          style={
            Object {
              "--iconsAfter": 0,
              "--iconsBefore": 0,
            }
          }
        >
          <span
            className="input"
          >
            <input
              className="input"
              id="postcode"
              name="postcode"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            />
          </span>
          <span
            className="before"
          />
          <span
            className="after"
          />
        </span>
        <p
          className="root"
        />
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        Cancel
      </span>
    </button>
    <button
      className="button"
      type="submit"
    >
      <span
        className="content"
      >
        Use Card
      </span>
    </button>
  </div>
</form>
`;
